File: bindings/swift/Sources/Sone/Nodes.swift
import Foundation

/// Collects children written as statements in a trailing closure — the same
/// shape SwiftUI uses, and what gives `if` and `for` inside the block.
@resultBuilder
public enum NodeBuilder {
    public static func buildBlock(_ components: [Node]...) -> [Node] { components.flatMap { $0 } }
    public static func buildExpression(_ expression: Node) -> [Node] { [expression] }
    public static func buildExpression(_ expression: [Node]) -> [Node] { expression }
    public static func buildOptional(_ component: [Node]?) -> [Node] { component ?? [] }
    public static func buildEither(first component: [Node]) -> [Node] { component }
    public static func buildEither(second component: [Node]) -> [Node] { component }
    public static func buildArray(_ components: [[Node]]) -> [Node] { components.flatMap { $0 } }
    public static func buildLimitedAvailability(_ component: [Node]) -> [Node] { component }
}

/// Collects paragraph content.
@resultBuilder
public enum InlineBuilder {
    public static func buildBlock(_ components: [Inline]...) -> [Inline] { components.flatMap { $0 } }
    public static func buildExpression(_ expression: Inline) -> [Inline] { [expression] }
    public static func buildExpression(_ expression: String) -> [Inline] { [.text(expression)] }
    public static func buildExpression(_ expression: Span) -> [Inline] { [.span(expression)] }
    public static func buildExpression(_ expression: [Inline]) -> [Inline] { expression }
    public static func buildOptional(_ component: [Inline]?) -> [Inline] { component ?? [] }
    public static func buildEither(first component: [Inline]) -> [Inline] { component }
    public static func buildEither(second component: [Inline]) -> [Inline] { component }
    public static func buildArray(_ components: [[Inline]]) -> [Inline] { components.flatMap { $0 } }
}

/// A vertical container.
public final class Column: Node, LayoutNode {
    public init(@NodeBuilder _ children: () -> [Node] = { [] }) {
        super.init(type: "column")
        self.children = children()
    }
}

/// A horizontal container.
public final class Row: Node, LayoutNode {
    public init(@NodeBuilder _ children: () -> [Node] = { [] }) {
        super.init(type: "row")
        self.children = children()
    }
}

/// A grid container with row-major auto placement.
public final class Grid: Node, LayoutNode {
    public init(@NodeBuilder _ children: () -> [Node] = { [] }) {
        super.init(type: "grid")
        self.children = children()
    }

    @discardableResult public func columns(_ tracks: Track...) -> Self { set("columns", .array(tracks.map { $0.ir })) }
    @discardableResult public func rows(_ tracks: Track...) -> Self { set("rows", .array(tracks.map { $0.ir })) }
    @discardableResult public func autoRows(_ tracks: Track...) -> Self { set("autoRows", .array(tracks.map { $0.ir })) }
    @discardableResult public func autoColumns(_ tracks: Track...) -> Self { set("autoColumns", .array(tracks.map { $0.ir })) }
}

/// A styled run inside a `Text`.
public final class Span: Node, SpanStyled {
    public init(_ text: String) {
        super.init(type: "span")
        inline = [.text(text)]
    }
}

/// A paragraph. Both a box and a run of text — Swift protocols compose all three
/// property sets with no single-inheritance fight.
public final class Text: Node, LayoutNode, SpanStyled, TextBlock {
    public init(_ text: String) {
        super.init(type: "text")
        inline = [.text(text)]
    }

    public init(@InlineBuilder _ content: () -> [Inline]) {
        super.init(type: "text")
        inline = content()
    }
}

/// Cascades text styling onto its descendants without drawing a box.
public final class TextDefault: Node, SpanStyled, TextBlock {
    public init(@NodeBuilder _ children: () -> [Node] = { [] }) {
        super.init(type: "text-default")
        self.children = children()
    }
}

/// An image.
public final class Photo: Node, LayoutNode {
    /// From a path, a URL, or `asset:name`.
    public init(_ src: String) {
        super.init(type: "photo")
        props.set("src", .string(src))
    }

    /// From raw bytes, inlined into the document as a data URL.
    public init(data: Data) {
        super.init(type: "photo")
        props.set("src", .string("data:application/octet-stream;base64," + data.base64EncodedString()))
    }

    /// How the image fills its box. The alignment is 0..1.
    @discardableResult public func scaleType(_ value: ScaleType, alignment: Double? = nil) -> Self {
        props.set("scaleType", .string(value.rawValue))
        if let alignment { props.set("scaleAlignment", .number(alignment)) }
        return self
    }

    @discardableResult public func preserveAspectRatio(_ value: Bool = true) -> Self { set("preserveAspectRatio", .boolean(value)) }
    @discardableResult public func flipHorizontal(_ value: Bool = true) -> Self { set("flipHorizontal", .boolean(value)) }
    @discardableResult public func flipVertical(_ value: Bool = true) -> Self { set("flipVertical", .boolean(value)) }

    /// The letterbox colour behind a `contain` image.
    @discardableResult public func fill(_ color: String) -> Self { set("fill", .string(color)) }

    /// An SVG path the image is clipped to.
    @discardableResult public func clipPath(_ path: String) -> Self { set("clipPath", .string(path)) }
}

/// An SVG path.
public final class SvgPath: Node, LayoutNode {
    public init(_ d: String) {
        super.init(type: "path")
        props.set("d", .string(d))
    }

    @discardableResult public func stroke(_ color: String) -> Self { set("stroke", .string(color)) }
    @discardableResult public func strokeWidth(_ value: Double) -> Self { set("strokeWidth", .number(value)) }
    @discardableResult public func strokeLineCap(_ value: StrokeCap) -> Self { set("strokeLineCap", .string(value.rawValue)) }
    @discardableResult public func strokeLineJoin(_ value: StrokeJoin) -> Self { set("strokeLineJoin", .string(value.rawValue)) }
    @discardableResult public func strokeMiterLimit(_ value: Double) -> Self { set("strokeMiterLimit", .number(value)) }
    @discardableResult public func strokeDashArray(_ values: Double...) -> Self { set("strokeDashArray", .array(values.map { .number($0) })) }
    @discardableResult public func strokeDashOffset(_ value: Double) -> Self { set("strokeDashOffset", .number(value)) }
    @discardableResult public func fill(_ color: String) -> Self { set("fill", .string(color)) }
    @discardableResult public func fillOpacity(_ value: Double) -> Self { set("fillOpacity", .number(value)) }
    @discardableResult public func fillRule(_ value: FillRule) -> Self { set("fillRule", .string(value.rawValue)) }

    /// Scale the path data itself, before layout.
    @discardableResult public func scalePath(_ value: Double) -> Self { set("scalePath", .number(value)) }
}

/// A table. Children are `TableRow`s.
public final class Table: Node, LayoutNode {
    public init(@NodeBuilder _ children: () -> [Node] = { [] }) {
        super.init(type: "table")
        self.children = children()
    }

    /// Row and column spacing. One argument sets both.
    @discardableResult public func spacing(_ row: Double, _ column: Double? = nil) -> Self {
        set("spacing", .array([.number(row), .number(column ?? row)]))
    }
}

/// A table row. Children are `TableCell`s.
public final class TableRow: Node, LayoutNode {
    public init(@NodeBuilder _ children: () -> [Node] = { [] }) {
        super.init(type: "table-row")
        self.children = children()
    }
}

/// A table cell.
public final class TableCell: Node, LayoutNode {
    public init(@NodeBuilder _ children: () -> [Node] = { [] }) {
        super.init(type: "table-cell")
        self.children = children()
    }

    @discardableResult public func colspan(_ value: Int) -> Self { set("colspan", .integer(value)) }
    @discardableResult public func rowspan(_ value: Int) -> Self { set("rowspan", .integer(value)) }
}

/// A bulleted or numbered list. Named `Bullets` because `List` is SwiftUI's.
public final class Bullets: Node, LayoutNode {
    public init(@NodeBuilder _ children: () -> [Node] = { [] }) {
        super.init(type: "list")
        self.children = children()
    }

    /// `disc`, `circle`, `square`, `decimal`, `dash`, `none`, or literal text.
    @discardableResult public func listStyle(_ value: String) -> Self { set("listStyle", .string(value)) }

    /// A styled marker node. `{}` in its text is replaced with the item number.
    @discardableResult public func listStyle(_ marker: Node) -> Self { set("listStyle", .node(marker)) }

    @discardableResult public func markerGap(_ value: Double) -> Self { set("markerGap", .number(value)) }
    @discardableResult public func markerOffset(_ value: Double) -> Self { set("markerOffset", .number(value)) }
    @discardableResult public func startIndex(_ value: Int) -> Self { set("startIndex", .integer(value)) }
}

/// One item in a `Bullets` list.
public final class ListItem: Node, LayoutNode {
    public init(@NodeBuilder _ children: () -> [Node] = { [] }) {
        super.init(type: "list-item")
        self.children = children()
    }

    /// Override the list's marker for this item alone.
    @discardableResult public func marker(_ value: Node) -> Self { set("marker", .node(value)) }
}

/// Clips every child to an SVG path.
public final class ClipGroup: Node, LayoutNode {
    public init(_ clipPath: String, @NodeBuilder _ children: () -> [Node] = { [] }) {
        super.init(type: "clip-group")
        props.set("clipPath", .string(clipPath))
        self.children = children()
    }
}

/// An explicit page break. Only meaningful with `pageHeight` set.
public func PageBreak() -> Column {
    Column().height(0).pageBreak(.before)
}

File: crates/sone/src/nodes.rs
//! The node types.
//!
//! One struct per IR node type, so the compiler can tell them apart. That is
//! what lets `Text::size` mean the font size while `Column::size` means the
//! box — the ambiguity every other binding had to resolve by hand.

use sone_core::ir::{self, Node, NodeType, NodeTypeTag};

use crate::IntoNode;

macro_rules! node_types {
    ($( $(#[$meta:meta])* $ty:ident => $variant:ident ),* $(,)?) => {
        $(
            $(#[$meta])*
            #[derive(Debug, Clone)]
            pub struct $ty {
                pub(crate) node: Node,
            }

            impl $ty {
                pub(crate) fn new() -> Self {
                    $ty {
                        node: Node {
                            ty: NodeTypeTag(NodeType::$variant),
                            ..Default::default()
                        },
                    }
                }
            }

            impl IntoNode for $ty {
                fn into_node(self) -> Node {
                    self.node
                }
            }

            impl From<$ty> for Node {
                fn from(value: $ty) -> Node {
                    value.node
                }
            }
        )*
    };
}

node_types! {
    /// A vertical container.
    Column => Column,
    /// A horizontal container.
    Row => Row,
    /// A grid container with row-major auto placement.
    Grid => Grid,
    /// A paragraph of text and styled runs.
    Text => Text,
    /// A styled run inside a [`Text`].
    Span => Span,
    /// Cascades text styling onto its descendants without drawing a box.
    TextDefault => TextDefault,
    /// An image.
    Photo => Photo,
    /// An SVG path.
    SvgPath => Path,
    /// A table. Children are rows.
    Table => Table,
    /// A table row. Children are cells.
    TableRow => TableRow,
    /// A table cell.
    TableCell => TableCell,
    /// A bulleted or numbered list. Children are items.
    Bullets => List,
    /// One item in a [`Bullets`] list.
    ListItem => ListItem,
    /// Clips every child to an SVG path.
    ClipGroup => ClipGroup,
}

impl_layout_props!(
    Column, Row, Grid, Text, Photo, SvgPath, Table, TableRow, TableCell, Bullets, ListItem,
    ClipGroup,
);
impl_box_size!(
    Column, Row, Grid, Photo, SvgPath, Table, TableRow, TableCell, Bullets, ListItem, ClipGroup,
);
impl_span_props!(Text, Span, TextDefault);
impl_text_block_props!(Text, TextDefault);
impl_children!(
    Column, Row, Grid, TextDefault, Table, TableRow, TableCell, Bullets, ListItem, ClipGroup,
);

impl Text {
    /// The **font** size, not the box size — matching the TypeScript API, where
    /// a text builder omits the layout `size`. Use `width` and `height` for the
    /// box; they are still here, because `Text` is a box too.
    pub fn size(mut self, value: impl crate::value::Num) -> Self {
        self.node.props.size = Some(value.as_f32());
        self
    }

    /// Append raw text.
    pub fn content(mut self, text: impl Into<String>) -> Self {
        self.node.inline.push(ir::Inline::Text(text.into()));
        self
    }

    /// Append a styled run.
    pub fn span(mut self, span: Span) -> Self {
        self.node.inline.push(ir::Inline::Span(span.into_node()));
        self
    }
}

impl Span {
    /// The font size.
    pub fn size(mut self, value: impl crate::value::Num) -> Self {
        self.node.props.size = Some(value.as_f32());
        self
    }

    /// Append raw text.
    pub fn content(mut self, text: impl Into<String>) -> Self {
        self.node.inline.push(ir::Inline::Text(text.into()));
        self
    }
}

impl TextDefault {
    /// The font size cascaded onto descendants.
    pub fn size(mut self, value: impl crate::value::Num) -> Self {
        self.node.props.size = Some(value.as_f32());
        self
    }
}

impl Grid {
    pub fn columns(mut self, tracks: impl IntoIterator<Item = ir::GridTrack>) -> Self {
        self.node.props.columns = Some(tracks.into_iter().collect());
        self
    }

    pub fn rows(mut self, tracks: impl IntoIterator<Item = ir::GridTrack>) -> Self {
        self.node.props.rows = Some(tracks.into_iter().collect());
        self
    }

    pub fn auto_rows(mut self, tracks: impl IntoIterator<Item = ir::GridTrack>) -> Self {
        self.node.props.auto_rows = Some(tracks.into_iter().collect());
        self
    }

    pub fn auto_columns(mut self, tracks: impl IntoIterator<Item = ir::GridTrack>) -> Self {
        self.node.props.auto_columns = Some(tracks.into_iter().collect());
        self
    }
}

impl Photo {
    /// How the image fills its box.
    pub fn scale_type(mut self, value: ir::ScaleType) -> Self {
        self.node.props.scale_type = Some(value);
        self
    }

    /// 0 is start, 0.5 centre, 1 end.
    pub fn scale_alignment(mut self, value: impl crate::value::Num) -> Self {
        self.node.props.scale_alignment = Some(value.as_f32());
        self
    }

    pub fn preserve_aspect_ratio(mut self, value: bool) -> Self {
        self.node.props.preserve_aspect_ratio = Some(value);
        self
    }

    pub fn flip_horizontal(mut self, value: bool) -> Self {
        self.node.props.flip_horizontal = Some(value);
        self
    }

    pub fn flip_vertical(mut self, value: bool) -> Self {
        self.node.props.flip_vertical = Some(value);
        self
    }

    /// The letterbox colour behind a `contain` image.
    pub fn fill(mut self, color: impl Into<String>) -> Self {
        self.node.props.fill = Some(color.into());
        self
    }

    /// An SVG path the image is clipped to.
    pub fn clip_path(mut self, path: impl Into<String>) -> Self {
        self.node.props.clip_path = Some(path.into());
        self
    }
}

impl SvgPath {
    pub fn stroke(mut self, color: impl Into<String>) -> Self {
        self.node.props.stroke = Some(color.into());
        self
    }

    pub fn stroke_width(mut self, value: impl crate::value::Num) -> Self {
        self.node.props.stroke_width = Some(value.as_f32());
        self
    }

    pub fn stroke_line_cap(mut self, value: ir::StrokeCap) -> Self {
        self.node.props.stroke_line_cap = Some(value);
        self
    }

    pub fn stroke_line_join(mut self, value: ir::StrokeJoin) -> Self {
        self.node.props.stroke_line_join = Some(value);
        self
    }

    pub fn stroke_miter_limit(mut self, value: impl crate::value::Num) -> Self {
        self.node.props.stroke_miter_limit = Some(value.as_f32());
        self
    }

    pub fn stroke_dash_array(mut self, values: impl IntoIterator<Item = f32>) -> Self {
        self.node.props.stroke_dash_array = Some(values.into_iter().collect());
        self
    }

    pub fn stroke_dash_offset(mut self, value: impl crate::value::Num) -> Self {
        self.node.props.stroke_dash_offset = Some(value.as_f32());
        self
    }

    pub fn fill(mut self, color: impl Into<String>) -> Self {
        self.node.props.fill = Some(color.into());
        self
    }

    pub fn fill_opacity(mut self, value: impl crate::value::Num) -> Self {
        self.node.props.fill_opacity = Some(value.as_f32());
        self
    }

    pub fn fill_rule(mut self, value: ir::FillRule) -> Self {
        self.node.props.fill_rule = Some(value);
        self
    }

    /// Scale the path data itself, before layout.
    pub fn scale_path(mut self, value: impl crate::value::Num) -> Self {
        self.node.props.scale_path = Some(value.as_f32());
        self
    }
}

impl Table {
    /// Row and column spacing.
    pub fn spacing(mut self, row: impl crate::value::Num, column: impl crate::value::Num) -> Self {
        self.node.props.spacing = Some(vec![row.as_f32(), column.as_f32()]);
        self
    }
}

impl TableCell {
    pub fn colspan(mut self, value: impl crate::value::Int) -> Self {
        self.node.props.colspan = Some(value.as_i64() as u32);
        self
    }

    pub fn rowspan(mut self, value: impl crate::value::Int) -> Self {
        self.node.props.rowspan = Some(value.as_i64() as u32);
        self
    }
}

impl Bullets {
    /// `"disc"`, `"decimal"`, literal text, or a styled marker node.
    pub fn list_style(mut self, value: impl Into<ir::ListStyle>) -> Self {
        self.node.props.list_style = Some(value.into());
        self
    }

    pub fn marker_gap(mut self, value: impl crate::value::Num) -> Self {
        self.node.props.marker_gap = Some(value.as_f32());
        self
    }

    pub fn marker_offset(mut self, value: impl crate::value::Num) -> Self {
        self.node.props.marker_offset = Some(value.as_f32());
        self
    }

    pub fn start_index(mut self, value: impl crate::value::Int) -> Self {
        self.node.props.start_index = Some(value.as_i64() as i32);
        self
    }
}

impl ListItem {
    /// Override the list's marker for this item alone.
    pub fn marker(mut self, marker: impl IntoNode) -> Self {
        self.node.props.marker = Some(Box::new(marker.into_node()));
        self
    }
}
